Sample Dom0 functionalities PR for internal review

firscity opened the pull request.

firscity committed:
xenvm: dom0: update Xen public headers for running Zephyr as Dom0

Xen public headers are added to Zephyr kernel partially, so we need
to update them for new features implementation.

Further implementation of Xen domains configuration and memory
management requires dedicated Xen public headers inside Zephyr kernel.
It will be used for add Zephyr OS Xen Domain-0 funtionalities.

Update existent and add new required public headers from Xen 4.17.0
release.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: change HVM functions signature to run it for other domains

This commit adds possibility to call hypervisor HVM parameter functions
for specified domain (instead of only DOMID_SELF). It is needed for
configuring domains, that were created from Zephyr control domain.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xenvm: events: add event channel allocation for domain-0

This commit adds interface for evtchn allocation that can be used by
privilaged domain. Domain 0 can specify both dom and remote_dom
parameters for hypercall, where in others domains dom should be always
DOMID_SELF. It is needed for creating pre-defined channels during
domain setup in Zephyr Dom0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: xenvm: move XEN_INITIAL_DOMAIN config to Xen driver

Untie XEN_INITIAL_DOMAIN from XenVM SoC configuration, it is
necessary for running Xen dom-0 on different boards. Move it from
XenVM SoC configuration to Xen driver. Set XEN_INITIAL_DOMAIN by
default to n in order to prevent issues with existing setups,
which may use XenVM as DomU.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: bindings: xen: add xen,xen.yaml file

Add yaml file for 'xen,xen', because without it an appropriate
'CONFIG_DT_HAS_XEN_XEN_ENABLED' isn't generated.

Remove xen,xen-4.15.yaml at all, because it isn't necessary to have
yaml for some specific Xen version.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use driver for Xen
Initial domain shield, that means that the driver isn't tied only
to XenVM board configuration.

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use arch arm64 core Xen
related code for Xen Initial domain shield, that means that the core
of Xen isn't tied only to XenVM SoC configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen

Unfortunately, set/way ops are not easily virtualized by Xen. S/W
emulation is disabled, because IP-MMU is active for Dom0. IP-MMU is
a IO-MMU made by Renesas, as any good IO-MMU, it shares page-tables
with CPU. Trying to emulate S/W with IP-MMU active will lead to
IO-MMU faults. So if we build Zephyr as a Xen Initial domain, it
won't work with cache management support enabled.

In this commit I use VA data invalidate instead of set/way --
invalidate all virtual address space, so it might increase system
initialization time.

Note: this patch is applied as a simple and fast workaround, we need
      to find more optimal solution.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: xen: gnttab: do Xen node mapping inside driver

Move memory mapping of Xen node to Grant Table driver system init
function. After moving mapping we don't need anymore records of
xen-xen node into 'mmu_regions' array, so they were deleted from
all SoCs: Rcar Gen3/Gen4 and XenVM.

We need at least 16M of virtual address space to map memory of Xen
node, so the virtual memory sized has been increased to 32 MB, it
should be enough for basic use-cases and mapping of 16M mem region
of Xen node.

Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB. And after the changes
is maps every 4KB page, so we need more XLAT tables.

Note: Xen Grant Table driver doesn't use Zephyr Device Model.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

firscity committed:
xen: implement Xen domain control interface

Add Xen domctl API implementation for Zephyr as control domain.

Previously Zephyr OS was used as unprivileged Xen domain (Domain-U),
but it also can be used as lightweight Xen control domain (Domain-0).
To implement such fuctionality additional Xen interfaces are needed.
One of them is Xen domain controls (domctls) - it allows to create,
configure and manage Xen domains.

Also, used it as a possibility to update files copyright and licenses
identifiers in touched files.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: add helper functions for Xen domain memory management

Add wrapper functions for Xen memory managment hypercall. They can be
used for unprivilaged Zephyr guest initialization and for domain
management, when Zephyr is used as Xen Domain 0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on boards/shields/xenvm_dom0/doc/index.rst:
```diff
@@ -0,0 +1,53 @@
+*********************************
+
+* add correct memory node, based on region Xen picked for Domain-0 on your setup;
+* add mapping of hypervisor node to appropriate SoC;
```
(note) looks like we don't need it anymore
@xakep-amatop FYI

firscity replied:
This node may be moved for different SoC, so this line is valid.

xakep-amatop replied:
yep, but MMU memory mapping is here meant under mapping, i.e. putting of the hypervisor node inside mmu_regions array, it isn't needed anymore, hypervisor node memory is mapped inside grant tables driver during initialization

but we need to add comment about adding hypervisor node for a board overlay inside this shield, as far as I see it has been deleted from common shield overlay

firscity replied:
Oh, I got it. Reworded.

xakep-amatop commented:
@firscity pls, fix this
```
 LONG_LINE: line length of 136 exceeds 100 columns
File:boards/shields/xenvm_dom0/example/xen.dts
Line:35
```

xakep-amatop commented on boards/shields/xenvm_dom0/doc/index.rst:
```diff
@@ -0,0 +1,53 @@
+Overlay for xenvm board, that is present in `board/` directory of this shield is QEMU Xen control
+domain example. To run such setup, you need to:
+
+* fetch and build Xen (e.g. RELEASE-4.17.0) for arm64 platform
```
:+1:

xakep-amatop commented on include/zephyr/xen/dom0/domctl.h:
```diff
@@ -0,0 +1,38 @@
+int xen_domctl_assign_dt_device(int domid, char *dtdev_path);
+int xen_domctl_bind_pt_irq(int domid, uint32_t machine_irq, uint8_t irq_type,
+		uint8_t bus, uint8_t device, uint8_t intx, uint8_t isa_irq,
+		uint16_t spi);
```
(note) we know that a line can have 100 symbols, so maybe we can avoid so ugly lines

xakep-amatop commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,64 @@
+
+int xendom_remove_from_physmap(int domid, xen_pfn_t gpfn)
+{
+	struct xen_remove_from_physmap xrfp = {0};
```
(note) in this file we also can use the next initializer:
```
struct xen_remove_from_physmap xrfp = {
	.domid = domid;
	.gpfn = gpfn;
};
```

xakep-amatop commented:
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`

dsemenets commented:
`Reviewed-by: Dmytro Semenets <dmytro_semenets@epam.com>`

firscity committed:
boards: shields: add virtual Xen dom0 shield support

Zephyr board shields allow to extend board features and services. It is
made by applying shield generic Kconfig/device-tree changes and
board-specific device-tree overlays. Originally it was intended to be
used for easier board prototyping and compatibility with applications.

Also it looks like great approach for convertion Zephyr OS build to Xen
control domain on different boards, as it requires Kconfig definition
and board device tree changes (configuration for memory node, hypervisor
console, grant table region etc).

Please note, that Xen hypervisor passes all selected parameters via
domain device-tree, but Zephyr does not use it in runtime. So, user
should keep values in board overlays in sync with real Xen values.

Add virtual Xen dom0 shield support and the dependence of config
UART_XEN_HVC_CONSOLEIO on configuration of shield 'xenvm_dom0', as this
way for console communication may be used as default only by privileged
Xen domain.

Add sample shield and docs for converting qemu_cortex_a53 board to Xen
Domain-0 guest. This approach might be used for other boards, where
Zephyr Dom0 is needed.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc commented on boards/shields/xenvm_dom0/boards/qemu_cortex_a53.overlay:
```diff
@@ -0,0 +1,43 @@
+/ {
+	chosen {
+		/delete-property/ zephyr,flash;
+		/delete-property/ zephyr,pcie-controller;
```
Why are you removing those two nodes?

firscity replied:
Added Xen dts in `example/` directory is pretty minimalistic, so I decided to make it as simple as possible and remove all redundant configuration from original qemu_cortex_a53 device tree.

I added comment above

lorc replied:
Hm, I believe, so it will be better other way around: show that full QEMU machine is supported in Dom0 mode. Otherwise, there is a little value in adding shield on top of it.

Consider this point of view: you are adding a Xen shield on top of working board, but you are disabling all the devices, changing memory sizes and entry addresses (probable).  What is left of the original board? :)

firscity replied:
@lorc, yes, you are right. In current version I:
- added complete node set to hypervisor sample device-tree and removed all deletion from overlay
- disabled UART instead of deleting
- added `dom0_mem` parameter
- updated memory node (region changed for 128MB)

lorc commented on boards/shields/xenvm_dom0/boards/qemu_cortex_a53.overlay:
```diff
@@ -0,0 +1,43 @@
+	/*
+	 * This node may differs on different setups, please check
+	 * following line in Xen boot log to set it right:
+	 * (XEN) Grant table range: 0x00000040200000-0x00000040240000
```
Please expand this comment by describing that Xen passes actual gntab ranges in device tree, but Zephyr is not capable of parsing device tree in run-time.

firscity replied:
Fixed

lorc commented on boards/shields/xenvm_dom0/boards/qemu_cortex_a53.overlay:
```diff
@@ -0,0 +1,43 @@
+	 * region for Domain-0 for every specific configuration. You can
+	 * start Xen for your platform and check following log:
+	 * (XEN) Allocating 1:1 mappings for dom0:
+	 * (XEN) BANK[0] 0x00000060000000-0x00000080000000
```
Please expand this comment by describing that Xen passes actual memory ranges in device tree, but Zephyr is not capable of parsing device tree in run-time.

firscity replied:
Fixed

lorc reviewed (commented):
I believe, you need to extend description of the last commit by telling why we are using shield.

Also, do not forget to remove `acked-by` tags

firscity reviewed (commented):
@lorc thanks for review

> I believe, you need to extend description of the last commit by telling why we are using shield.

I added few paragraphs about this approach for Dom0 to commit message.

> Also, do not forget to remove acked-by tags

Yes, sure, I will take commits without tags to mainline and freshen tags to our future `zephyr-v3.3.0-xt` branch

lorc reviewed (approved):
Thank you for the fixes. I am fine the patches in general. Probably there will be issues with that workaround patch, but anyways, it is good start,.

I am not adding formal tag, because this PR is intended to be upstreamed.

lorc reviewed (approved):
Thank you for the fixes. I am fine the patches in general. Probably there will be issues with that workaround patch, but anyways, it is good start,.

I am not adding formal tag, because this PR is intended to be upstreamed.

firscity commented:
@lorc thank you for a close review.

xakep-amatop commented:
Hi, @firscity 

Do we need this pull ?
as far as know we have full version of changes here https://github.com/xen-troops/zephyr/pull/43
